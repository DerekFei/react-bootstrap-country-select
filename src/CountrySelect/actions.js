import {
  MOUNT,
  FOCUS,
  BLUR,
  TEXT_CHANGE,
  ACTIVE_LIST_ITEM_CHANGE,
  TEXT_COMPLETION,
  COUNTRY_SELECT,
  CLEAR,
} from './constants';

export const mount = dispatch => (filterOptions, countries) => dispatch({
  type: MOUNT,
  filterOptions,
  countries,
});

export const focus = dispatch => () => dispatch({
  type: FOCUS,
});

export const blur = dispatch => () => dispatch({
  type: BLUR,
});

export const textChange = dispatch => inputText => dispatch({
  type: TEXT_CHANGE,
  inputText,
});

export const activeListItemChange = dispatch => activeListItemIndex => dispatch({
  type: ACTIVE_LIST_ITEM_CHANGE,
  activeListItemIndex,
});

export const textCompletion = dispatch => () => dispatch({
  type: TEXT_COMPLETION,
});

export const countrySelect = dispatch => selectedCountry => dispatch({
  type: COUNTRY_SELECT,
  selectedCountry,
});

export const clear = dispatch => () => dispatch({
  type: CLEAR,
});
