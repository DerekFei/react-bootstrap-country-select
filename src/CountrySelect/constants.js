export const MOUNT = 'MOUNT';
export const FOCUS = 'FOCUS';
export const BLUR = 'BLUR';
export const TEXT_CHANGE = 'TEXT_CHANGE';
export const ACTIVE_LIST_ITEM_CHANGE = 'ACTIVE_LIST_ITEM_CHANGE';
export const TEXT_COMPLETION = 'TEXT_COMPLETION';
export const COUNTRY_SELECT = 'COUNTRY_SELECT';
export const CLEAR = 'CLEAR';
