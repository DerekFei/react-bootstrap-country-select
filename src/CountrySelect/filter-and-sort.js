export const filter = (matchNameFromStart, matchAbbreviations) => inputText => {

  const regex = new RegExp(`${matchNameFromStart ? '^' : ''}${inputText}`, 'i');

  return matchAbbreviations
    ? ({ name, alpha2, alpha3, ioc }) => regex.test(name) || regex.test(alpha2) || regex.test(alpha3) || regex.test(ioc)
    : ({ name }) => regex.test(name);

};

export const sort = sortBy => (a, b) => a[sortBy] < b[sortBy] ? -1 : 1;
