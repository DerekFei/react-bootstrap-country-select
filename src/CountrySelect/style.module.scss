.country-select {

  &:not(.flush) {

    .input-group__flag {
  
      min-width: 2.75em;
  
    }

    // &.sm {

    //   .input-group__flag {
  
    //     min-width: 3.75em;
    //     font-size: 0.625rem;
    
    //   }

    // }

  }

}
