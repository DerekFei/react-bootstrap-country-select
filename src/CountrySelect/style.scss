@import './variables.scss';

.country-select {

  &:not(.country-select--flush) {

    .country-select__input-group__flag {
  
      min-width: $input-group-flag-min-width;
  
    }

  }

}
