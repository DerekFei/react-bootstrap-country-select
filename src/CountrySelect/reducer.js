import { filter } from './filter-and-sort';

import {
  MOUNT,
  FOCUS,
  BLUR,
  TEXT_CHANGE,
  ACTIVE_LIST_ITEM_CHANGE,
  TEXT_COMPLETION,
  COUNTRY_SELECT,
  CLEAR,
} from './constants';

export const INITIAL_STATE = {
  focused: false,
  filterOptions: null,
  countries: [],
  inputText: '',
  list: [],
  activeListItemIndex: -1,
  activeCountry: null,
};

const mount = (state, { filterOptions, countries }) => ({
  ...state,
  filterOptions,
  countries,
  list: [ ...countries ],
});

const focus = state => ({
  ...state,
  focused: true,
});

const blur = state => ({
  ...state,
  focused: false,
});

const textChange = (state, { inputText }) => {

  const {
    filterOptions: {
      matchNameFromStart,
      matchAbbreviations,
    },
    countries,
    list,
    activeListItemIndex,
  } = state;

  const currentActiveCountry = list[activeListItemIndex];

  const countryFilter = filter(matchNameFromStart, matchAbbreviations)(inputText);

  const updatedList = [ ...countries.filter(countryFilter) ];

  let newActiveCountryIndex = -1;

  if (updatedList.length === 1) {

    newActiveCountryIndex = 0;

  } else if (currentActiveCountry) {

    const index = updatedList.findIndex(country => country.id === currentActiveCountry.id);

    newActiveCountryIndex = Number.isFinite(index) ? index : -1;

  }

  return {
    ...state,
    inputText,
    list: updatedList,
    activeListItemIndex: newActiveCountryIndex,
  };

};

const activeListItemChange = (state, { activeListItemIndex }) => ({
  ...state,
  activeListItemIndex,
});

const textCompletion = state => {

  const {
    inputText,
    list,
    activeListItemIndex,
  } = state;

  const activeCountry = list[activeListItemIndex];

  if (activeCountry && inputText.length < activeCountry.name.length) {

    const regex = new RegExp(`^(.{${inputText.length},${inputText.length}}\\s?[^\\s]+)`, 'i');

    const updatedInputText = activeCountry.name.match(regex)[1];

    return textChange(state, { inputText: updatedInputText });

  } else {

    return state;

  }

};

const countrySelect = (state, { selectedCountry }) => ({
  ...state,
  inputText: selectedCountry.name,
  list: [],
  activeListItemIndex: -1,
});

const clear = state => ({
  ...state,
  inputText: '',
  list: [ ...state.countries ],
  activeListItemIndex: -1,
});

const reducer = (state, action) => {
  switch (action.type) {
    case MOUNT: return mount(state, action);
    case FOCUS: return focus(state);
    case BLUR: return blur(state);
    case TEXT_CHANGE: return textChange(state, action);
    case ACTIVE_LIST_ITEM_CHANGE: return activeListItemChange(state, action);
    case TEXT_COMPLETION: return textCompletion(state, action);
    case COUNTRY_SELECT: return countrySelect(state, action);
    case CLEAR: return clear(state);
    default: throw new Error(`No such reducer action: ${action.type}`);
  }
};

export default reducer;
