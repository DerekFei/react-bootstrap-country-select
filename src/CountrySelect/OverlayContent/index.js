import React, { useRef } from 'react';

import List from './List';

import './style.scss';

const OverlayContent = ({
  classPrefix,
  list,
  activeListItemIndex,
  countryLabelFormatter,
  flags,
  noMatchesText,
  maxHeight = 225,
  onListItemClick,
}) => {

  const el = useRef(null);

  const handleAciveItemCutOff = overflowAmount => {

    el.current.scrollTop = el.current.scrollTop + overflowAmount;

  };

  return (
    <div
      className={`${classPrefix}__overlay-content`}
      style={{
        maxHeight: `${maxHeight}px`,
      }}
      ref={el}
    >

      {!(list.length >= 1) &&
        <div
          className={`${classPrefix}__no-matches`}
        >

          {noMatchesText}

        </div>
      }

      {(list.length >= 1) &&
        <List
          classPrefix={classPrefix}
          containerEl={el.current}
          list={list}
          activeItemIndex={activeListItemIndex}
          countryLabelFormatter={countryLabelFormatter}
          flags={flags}
          onItemClick={onListItemClick}
          onActiveItemOverflow={handleAciveItemCutOff}
        />
      }
    </div>
  );

};

export default OverlayContent;
