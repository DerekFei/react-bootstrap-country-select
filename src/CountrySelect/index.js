import React, { useState, useRef, useEffect, useReducer } from 'react';
import PropTypes from 'prop-types';
import { InputGroup, FormControl, Overlay } from 'react-bootstrap';

import { COUNTRIES } from './data';

import {
  applyExclusionsAndAdditions,
  removeEmojiFlag,
  getUpdatedList,
} from './util';

import OverlayContent from './OverlayContent';

import reducer, { INITIAL_STATE } from './reducer';

import {
  init,
  focus,
  blur,
  textChange,
  activeListItemChange,
  countrySelect,
  clear,
} from './actions';

import style from './style.module.scss';

const CountrySelect = ({
  value,
  onChange = () => {},
  onTextChange,
  countries = COUNTRIES,
  exclusions,
  additions,
  valueAs = 'object', // 'object' | 'id'
  flags = true,
  flush = true,
  disabled = false,
  placeholder = 'Type or select country...',
  noMatchesText = 'No matches',
  size, // 'sm' | 'lg'
  sort, // e.g. (c1, c2) => c1.name < c2.name ? -1 : (c1.name > c2.name ? 1 : 0),
  matchNameFromStart = true,
  matchAbbreviations = false,
  countryLabelFormatter = ({ name }) => name,
  throwInvalidValueError = false, 
  listMaxHeight,
  formControlProps = {},
  overlayProps = {},
}) => {

  const inputGroupRef = useRef(null);
  const formControlRef = useRef(null);
  const hasInitRef = useRef(false);
  const [ width, setWidth ] = useState(-1);

  const [ {
    focused,
    inputText,
    list,
    activeListItemIndex,
    combinedCountries,
  }, dispatch ] = useReducer(reducer, INITIAL_STATE);

  const handleFocus = focus(dispatch);
  const handleBlur = blur(dispatch);
  const handleTextChange = textChange(dispatch);
  const handleListActiveItemChange = activeListItemChange(dispatch);
  const handleCountrySelect = countrySelect(dispatch);
  const handleClear = clear(dispatch);

  const selectedCountry = value ? (combinedCountries || []).find(country => country.id === (value.id || value)) : null;

  if (throwInvalidValueError && value && !selectedCountry)
    throw new Error(`No matching country for value: ${JSON.stringify(value)}`);

  useEffect(() => {

    if (hasInitRef.current) return;

    const combinedCountries = applyExclusionsAndAdditions(countries, exclusions, additions);

    init(dispatch)(combinedCountries);

    hasInitRef.current = true;

  }, [ countries, exclusions, additions ]);

  useEffect(() => {

    setWidth(inputGroupRef.current.offsetWidth);

  }, [ inputGroupRef ]);

  const select = listItemIndex => {

    const country = list[listItemIndex];

    handleCountrySelect();
    onChange(valueAs === 'id' ? country.id : country);

  };

  const escape = () => {

    handleClear();
    onChange(null);

  };

  const inputChange = text => {

    if (selectedCountry && flags) {

      text = removeEmojiFlag(text);

    }

    const [ updatedList, updatedActiveListItemIndex ]
      = getUpdatedList(text, list, activeListItemIndex, combinedCountries, sort, matchNameFromStart, matchAbbreviations);

    handleTextChange(text, updatedList, updatedActiveListItemIndex);

    if (onTextChange) onTextChange(text);
    if (value) onChange(null);

  };

  const handleKey = ev => {

    if (ev.key === 'ArrowUp') {

      ev.preventDefault();

      const newIndex = activeListItemIndex <= 0 ? list.length - 1 : activeListItemIndex - 1;
      handleListActiveItemChange(newIndex);

    } else if (ev.key === 'ArrowDown') {

      const newIndex = activeListItemIndex >= list.length - 1 ? 0 : activeListItemIndex + 1;
      handleListActiveItemChange(newIndex);

    } else if (ev.key === 'Enter') {

      if (activeListItemIndex >= 0) select(activeListItemIndex)

    } else if (ev.key === 'Escape') {

      escape();

    }

  };

  return (
    <div className={`${style['country-select']} ${flush ? style.flush : ''} ${size ? style[size] : ''} ${focused ? style.focused : ''}`}>

      <InputGroup
        ref={inputGroupRef}
        className={style['input-group']}
        size={size}
      >

        { (!flush && flags) && 
          <InputGroup.Prepend>

            <InputGroup.Text
              className={style['input-group__flag']}
            >

              {selectedCountry ? selectedCountry.flag : ''}
            
            </InputGroup.Text>
            
          </InputGroup.Prepend>
        }

        <FormControl
          ref={formControlRef}
          className={style['form-control']}
          value={selectedCountry ? `${flush && flags ? selectedCountry.flag + '  ' : ''}${selectedCountry.name}` : inputText}
          onKeyDown={handleKey}
          onChange={ev => inputChange(ev.target.value)}
          onFocus={handleFocus}
          onBlur={handleBlur}
          placeholder={placeholder}
          disabled={disabled}
          spellCheck={false}
          autoComplete='false'
          {...formControlProps}
        />

      </InputGroup>

      <Overlay
        target={inputGroupRef.current}
        rootClose
        placement='bottom-start'
        show={focused && !selectedCountry}
        onHide={() => {}}
        transition
        {...overlayProps}
      >

        {({ placement, arrowProps, show: _show, popper, ...props }) => (

          <div
            {...props}
            style={{
              width: (width > 0) ? `${width}px` : 'calc(100% - 10px)',
              ...props.style,
            }}
          >

            <OverlayContent
              list={list}
              activeListItemIndex={activeListItemIndex}
              countryLabelFormatter={countryLabelFormatter}
              flags={flags}
              noMatchesText={noMatchesText}
              maxHeight={listMaxHeight}
              onListItemClick={select}
            />

          </div>

        )}

      </Overlay>
    
    </div>
  );

};

CountrySelect.propTypes = {
  value: PropTypes.oneOfType([
    PropTypes.shape({
      id: PropTypes.string.isRequired,
    }),
    PropTypes.string,
  ]),
  onChange: PropTypes.func,
  onTextChange: PropTypes.func,
  countries: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
  })),
  exclusions: PropTypes.arrayOf(PropTypes.string),
  additions: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
  })),
  valueAs: PropTypes.oneOf([ 'object', 'id' ]),
  flags: PropTypes.bool,
  flush: PropTypes.bool,
  disabled: PropTypes.bool,
  placeholder: PropTypes.node,
  noMatchesText: PropTypes.node,
  size: PropTypes.oneOf([ 'sm', 'lg' ]),
  sort: PropTypes.func,
  matchNameFromStart: PropTypes.bool,
  matchAbbreviations: PropTypes.bool,
  countryLabelFormatter: PropTypes.func,
  throwInvalidValueError: PropTypes.bool, 
  listMaxHeight: PropTypes.number,
  formControlProps: PropTypes.object,
  overlayProps: PropTypes.object,
};

export default CountrySelect;
