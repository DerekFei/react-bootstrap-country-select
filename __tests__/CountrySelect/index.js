import React from 'react';
import renderer from 'react-test-renderer'

import CountrySelect from '../../src/CountrySelect/index.js';

describe('CountrySelect', () => {

  it('should render correctly without props', () => {

    const tree = renderer
      .create(
        <CountrySelect
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();

  });

});
