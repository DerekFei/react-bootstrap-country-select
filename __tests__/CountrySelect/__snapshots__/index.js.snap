// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountrySelect should render correctly without props 1`] = `
<div
  className="country-select country-select--flush"
>
  <div
    className="country-select__input-group input-group"
  >
    <input
      autoComplete="false"
      className="country-select__form-control form-control"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="Type or select country..."
      spellCheck={false}
      value=""
    />
  </div>
</div>
`;
